### ISSUE TITLE
Loading Qt 5 plugins does not work out of the box anymore

### ISSUE THREAD

**Issue by xantares** [Martchus/PKGBUILDs#109]
it seems all qt executables are broken:
```
x86_64-w64-mingw32-wine  /usr/x86_64-w64-mingw32/bin/qcollectiongenerator-qt5.exe -v
```
fails because of unimplemented RaiseFailFastException in wine

it comes from 
https://github.com/qt/qtbase/blob/5b193e3dd47704daf56c8817d0157f66363044c3/src/corelib/global/qlogging.cpp#L1885

but even if I apply patch from https://bugs.winehq.org/show_bug.cgi?id=46155 to implement the function (and do nothing inside)
I get "returned a non-zero code: 21"

any ideas ?

**Comment by Martchus** [comment]
So far I don't have an idea. The apps I've been building with Qt 5.14 from my repo are working as far as I know (e.g. latest release of [syncthingtray](https://github.com/Martchus/syncthingtray/releases) and [tagedtior](https://github.com/Martchus/tageditor/releases)).

Ignoring the issue title it seems that you mean the problem would only occur if one uses WINE and if `qt_message_fatal` is called somewhere by the app. So I guess this is actually only about this particular use-case? And it doesn't seem that the implementation of the function from the bug you've mentioned does nothing at all (so it might not be correct).

By the way, why are you invoking `qcollectiongenerator-qt5.exe`? Is this just an example? I found it always weird that this tool isn't a host tool (especially when considering "casual cross targets" like Android/iOS - would you really want to generate Qt's help on your phone?).

**Comment by xantares** [comment]
Sorry, I didnt mean apps, but just the qt tools from, ie those in /usr/*-w64-mingw32/bin/qt*.exe

**Comment by Martchus** [comment]
By the way, since the program is supposed to terminate at this point anyways, I'd just ignore this type of error as a workaround. But apparently you're running this under Linux (why else would you be using WINE). So maybe you can just use the native `qcollectiongenerator` executable.

**Comment by xantares** [comment]
> By the way, since the program is supposed to terminate at this point anyways, I'd just ignore this type of error as a workaround. But apparently you're running this under Linux (why else would you be using WINE). So maybe you can just use the native `qcollectiongenerator` executable.
> 
I think that may be a bug of since the native "qhelpgenerator -v" prints the version number and exit nicely
the previous mingw-w64-qt 5.13.2 binaries did not have that problem

> Oh, and the reason why `qcollectiongenerator-qt5.exe` might not be a host application although it would make more sense is likely that it depends on stuff not present in the bootstrapping library.
Yes, that's it, and not depending on the native qt libs save some space.

**Comment by Martchus** [comment]
These apps are likely not supposed to log fatal error messages in the first place - so that sounds like a regression. What does the error message say, by the way? And which tools are affected exactly? And this is only happening when starting them with WINE, right? I can do my own tests later this evening.

**Comment by xantares** [comment]
Now I see the error popup: 
"this application failed to start because no qt platform plugin could be initialized"
you can see it if you run:
```
x86_64-w64-mingw32-wine /usr/x86_64-w64-mingw32/bin/qhelpgenerator-qt5.exe -v
```
I tried to set QT_PLUGIN_PATH=/usr/x86_64-w64-mingw32/lib/qt/plugins or QT_DEBUG_PLUGINS without luck

**Comment by Martchus** [comment]
I've just tried to reproduce the problem but I can not:

```
[martchus@martchus-arch chroot]$ x86_64-w64-mingw32-wine /usr/x86_64-w64-mingw32/bin/qhelpgenerator-qt5.exe -v
Qt Help Generator version 1.0 (Qt 5.14.0)
[martchus@martchus-arch chroot]$ i686-w64-mingw32-wine /usr/i686-w64-mingw32/bin/qhelpgenerator-qt5.exe -v
Qt Help Generator version 1.0 (Qt 5.14.0)
[martchus@martchus-arch chroot]$ i686-w64-mingw32-wine /usr/i686-w64-mingw32/bin/qhelpgenerator-qt5.exe 
Missing input file name.


Usage:

qhelpgenerator <file> [options]

  -o <output-file>       Generates a Qt compressed help
                         called <output-file> (*.qch) for the
                         Qt help project <file> (*.qhp).
                         Generates a Qt help collection
                         called <output-file> (*.qhc) for the
                         Qt help collection project <file> (*.qhcp).
                         If this option is not specified
                         a default name will be used
                         (*.qch for *.qhp and *.qhc for *.qhcp).
  -c                     Checks whether all links in HTML files
                         point to files in this help project.
  -s                     Suppresses status messages.
  -v                     Displays the version of 
                         qhelpgenerator.

[martchus@martchus-arch chroot]$ x86_64-w64-mingw32-wine /usr/x86_64-w64-mingw32/bin/qdbusviewer-qt5.exe 
002b:fixme:win:RegisterPowerSettingNotification (0x20052,{02731015-4510-4526-99e6-e5a17ebd1aea},0): stub
002b:fixme:ntdll:NtQuerySystemInformation info_class SYSTEM_PERFORMANCE_INFORMATION
002b:fixme:dwmapi:DwmIsCompositionEnabled 000000000032BFD0
002b:fixme:win:UnregisterPowerSettingNotification (0xdeadbeef): stub
[martchus@martchus-arch chroot]$
```

So also the graphical `qdbusviewer-qt5.exe` works. I'm using `mingw-w64-wine 1-5` but have *not* `mingw-w64-wine-qt 1-5` installed. It also works to use WINE directly with a clean `WINEPATH`, e.g. `WINEPATH= wine /usr/x86_64-w64-mingw32/bin/qdbusviewer-qt5.exe`. That's actually no surprise because the DLLs lie beside the exe. However, I'm still wondering how it is able to locate the plugins in my case. Maybe Qt's plugin system just looks under `lib/qt/plugins` by default? At least the string `lib/qt/plugins` is part of `Qt5Core.dll`.

I've recently tested my application `syncthingctl.exe` which is using `QFile` with WINE and noticed that it didn't like "native" paths like `/home/martchus/.config/…`. Instead I had to use `K:\home\martchus\.config\…`. Maybe that's also why your attempt to set `QT_PLUGIN_PATH` didn't help.

**Comment by xantares** [comment]
oh, that's not right. Maybe a wine bug ? I have 5.0 what's your version ?

**Comment by Martchus** [comment]
I'll try to test this in a clean chroot when I have time. And I've been using the latest WINE and the latest binaries from my repository.

**Comment by Martchus** [comment]
Ok, I can reproduce it in a clean chroot as well. But I also have no clue to make it work so far. The usual like setting `QT_DEBUG_PLUGINS` doesn't work (like you've already mentioned). But the apps work at least natively under Windows, right? Note that the statically linked apps work under WINE because they don't need to load any plugins. So it is really only loading the platform plugins which ceased to work. Maybe they did some changes about loading plugins in Qt 5.14? I'm still wondering why it works on my regular system. There must be some *magic* which makes it work.

**Comment by Martchus** [comment]
Note that there were definitely some changes in Qt itself: https://github.com/qt/qtbase/commits/5.14.0/src/corelib/plugin

**Comment by xantares** [comment]
Here the problem shows up on my regular machine too (it's up to date), maybe it comes from some core package that has not been updated yet on your regular machine ?

**Comment by Martchus** [comment]
I don't think so. My regular machine and the systemd container I've been using are at the latest Arch Linux testing.

Out of curiosity I've just did one more test within the systemd container:

```
[root@martchus /]# x86_64-w64-mingw32-wine /usr/x86_64-w64-mingw32/lib/qt/plugins/qhelpgenerator-qt5.exe -v
002a:err:module:load_so_dll failed to load .so lib "/usr/bin/../lib/wine/l3codeca.acm.so": libmpg123.so.0: cannot open shared object file: No such file or directory
Qt Help Generator version 1.0 (Qt 5.14.0)
```

So I've copied the `exe` file into the plugin directory so the plugins are located next to it similar as it would be when one actually installs this under Windows. That seems to work. It also works to specify the plugin path via `qt.conf`:

```
[root@martchus /]# cat /usr/x86_64-w64-mingw32/bin/qt.conf
[Paths]
Plugins = /usr/x86_64-w64-mingw32/lib/qt/plugins
[root@martchus /]# x86_64-w64-mingw32-wine /usr/x86_64-w64-mingw32/bin/qhelpgenerator-qt5.exe -v
002a:err:module:load_so_dll failed to load .so lib "/usr/bin/../lib/wine/l3codeca.acm.so": libmpg123.so.0: cannot open shared object file: No such file or directory
Qt Help Generator version 1.0 (Qt 5.14.0)
```

To summarize:

1. It is generally possible to load plugins.
2. `QT_DEBUG_PLUGINS` has no effect
3. `QT_PLUGIN_PATH` has no effect (regardless how you specify the path)
4. Qt is unable to locate the plugins in the way they are installed by our packaging. Since you're saying it worked with Qt 5.13 that's a regression.

Note that passing environment variables in the way we do works for sure, e.g. `ENABLE_ESCAPE_CODES=0 x86_64-w64-mingw32-wine syncthingctl.exe --help` really doesn't use escape codes. 

As a workaround I can include a `qt.conf` within `mingw-w64-qt5-base`. I'll have to update to 5.14.1 anyways.

**Comment by xantares** [comment]
How does this work since qt 5.14 is not supposed to find plugins in $PWD anymore:
https://github.com/qt/qtbase/commit/bf131e8d2181b3404f5293546ed390999f760404#diff-26102fab9c3c54dc01f971c39af049a3

And more importantly, how would we ship apps that rely on platform plugins if QT_PLUGIN_PATH is not taken into account ?

**Comment by Martchus** [comment]
> How does this work since qt 5.14 is not supposed to find plugins in $PWD anymore ?

The plugins are loaded from the directory where the executable is *stored*, that's not necessarily `$PWD`. This is also explained in the [documentation of QCoreApplication::libraryPaths()](https://doc.qt.io/qt-5/qcoreapplication.html#libraryPaths). So the commit makes actually sense because it makes *no* sense to randomly load DLLs from the current working directory.

> And more importantly, how would we ship apps that rely on platform plugins if QT_PLUGIN_PATH is not taken into account ?

I've already answered that:

> So I've copied the exe file into the plugin directory so the plugins are located next to it **similar as it would be when one actually installs this under Windows**.

But of course it is strange that `QT_PLUGIN_PATH` is ignored. So far I've only tested with WINE and I suspect that setting `QT_PLUGIN_PATH` works natively under Windows.

**Comment by Martchus** [comment]
I've been updating Qt to 5.14.1 and included `qt.conf` within the `bin` directory to workaround the issue.

**Comment by xantares** [comment]
thanks a lot

**Comment by Martchus** [comment]
I suppose I can close this issue now that the workaround is in place.
